### PR TITLE
Make `check` recognise `main-is` in conditional branches

### LINKED ISSUE #9742: Regression on master: `main-is` not picked up from conditional
On today's master ( 9025af597b905f2027a878c0aa0f34b885e42a6e ).
Example:
```cabal
executable twee
  if flag(parallel)
    main-is: ParallelMain.hs
    build-depends: async, unix
    c-sources: executable/link.c
  else
    main-is: Main.hs
```
(Source: https://github.com/nick8325/twee/blob/eb4b83b4df6dd16fbff091b3fd90b5a55c88d41a/twee.cabal#L46-L52)
`cabal build` errors:
```
Configuring executable 'twee' for twee-2.4.2...
Error: [Cabal-5559]
[no-main-is] No 'main-is' field found for executable twee
```
This works fine on 3.10.2.1, so must be a regression on `master`.
Related:
- #9724 : reports that `cabal sdist` behaves differently on master than in 3.10.2 for flag-selected `main-is`.

### PR DESCRIPTION
**Pull request by ffaf1**
Closes #9742.

Include the following checklist in your PR:

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#other-conventions).
* [ ] Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog).
* [x] The documentation has been updated, if necessary.
* [ ] [Manual QA notes](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#qa-notes) have been included.
* [x] Tests have been added. (*Ask for help if you don’t know how to write them! Ask for an exemption if tests are too complex for too little coverage!*)

### CODE REVIEW

**Review by geekosaur — APPROVED** [review]

**Comment by ffaf1** [comment]
I am wondering: does it makes sense to add a changelog? This patch fixes a bug of the reimplementation, and the reimplementation will land in the yet-to-be-released `3.12`. A changelog would only add noise, am I wrong?

**Comment by ulysses4ever** [comment]
If the currently released version doesn't have the bug, the changelog about fixing it is not needed imo.

File: Cabal/src/Distribution/PackageDescription/Check/Conditional.hs [hunk 1]
```diff
@@ -107,12 +108,55 @@ annotateCondTree fs ta (CondNode a c bs) =
...
+        fmap snd ms
+
+    utaBranch
+      :: a
+      -> CondBranch ConfVar [Dependency] (TargetAnnotation a)
+      -> CondBranch ConfVar [Dependency] (TargetAnnotation a)
+    utaBranch a (CondBranch k t mt) =
+      let utt (CondNode ka c wbs) =
+            (CondNode (updateTargetAnnotation a ka) c wbs)
+       in CondBranch k (utt t) (utt <$> mt)
```
Comment (Kleidukos) [inline Cabal/src/Distribution/PackageDescription/Check/Conditional.hs:152]: I'm not familiar with this part of the codebase. What do `rbf`,  `utt` and `uta` mean?

Comment (ffaf1) [reply Cabal/src/Distribution/PackageDescription/Check/Conditional.hs:152]: realiseBranchFunction, updateTargetAnnotation, updateTargetTree.

There is actually no need to have them as across, I will convert them to longhand.

**Comment by ffaf1** [comment]
@mergify backport 3.12

**Pull request merged**